Object extension

Skyhigh173 opened the pull request:
# blocks
![image](https://github.com/TurboWarp/extensions/assets/100467674/eeb10286-2c84-409d-99f1-21709dcc78cf)

> [!NOTE]
> You might think that JSON extension is enough, however Object extension is **way more faster** than json extension, as data will not be reparsed.
> All variables **will be converted to JSON string** when you save the project.

# performance/benchmark[^1]
- a ~8.4MB file:
![image](https://github.com/TurboWarp/extensions/assets/100467674/9fd0dc51-8419-4061-a860-5cc9f3e6cdf9)
![image](https://github.com/TurboWarp/extensions/assets/100467674/bc7af126-dd9f-4712-93b5-268e578d9fe0)

- a json string with length of 200
![image](https://github.com/TurboWarp/extensions/assets/100467674/381d867f-e90d-4768-82df-4ba299a0305e)
15x faster.

# Miscellaneous

(Read `IMPORTANT` part)

> [!IMPORTANT]
> it will be slower to parse/stringify the entire object due to reviver & replacer.

> [!IMPORTANT]
> variable monitor will not work properly, eg it shows `keyvalue` instead of `{"key":"value"}`

> [!IMPORTANT]
> use Cast Extension or `() to JSON` block to convert it to string.

> [!WARNING]
> this extension overwrites `Map.prototype.toString`.
> `Map._e` (extra utils) were added.

<details>
<summary>for developers (Map._e extra utils)</summary>

- use `Map._e.deepCopy` when you need to modify the content of the `Map`.
- use `Map._e.shallowCopy` when you only needs to read the content of the `Map`.
- use `Map._e.fromMapOrString` in set content related blocks.[^2]
- use `Map._e.fromArray` when you need to convert array to `Map`.
</details>

[^1]: Parsing time not counted.
[^2]: This function tries convert a string to `Map` if possible.

Skyhigh173 committed:
Create object.js

Skyhigh173 committed:
update to support array

Skyhigh173 committed:
remove test script

Skyhigh173 commented:
idk how to fix `Validate / format` part

snowboyz0825 commented:
into the archive it goes

CubesterYT commented:
> idk how to fix `Validate / format` part

@Skyhigh173 garbo can do it for you, mark this as ready, not draft.

LilyMakesThings commented:
> into the archive it goes

I can see a pattern here; please don't comment this on every single PR that is made, it's spam

Skyhigh173 commented:
> into the archive it goes

Wdym?

Skyhigh173 commented:
Planned thumbnail (subject to change):
![image](https://github.com/TurboWarp/extensions/assets/100467674/e209cefd-e3b3-486a-a004-8675286cdde8)

Skyhigh173 commented:
A bad news, you can't use SF symbols

do i have to find a new one 😅

Skyhigh173 committed:
Update - more blocks & bug fix

Skyhigh173 commented:
My project need this, so pls review it faster (if possible)

SharkPool-SP commented:
[Objects.txt](https://github.com/TurboWarp/extensions/files/13051450/Objects.txt)

I added 2 new blocks, they are incredibly helpful to me, and what I believe others would find useful too:
<img width="361" alt="Screenshot 2023-10-20 at 12 35 44 AM" src="https://github.com/TurboWarp/extensions/assets/139097378/9b91315e-59b6-44ce-982e-3be6442f1104">
<img width="361" alt="Screenshot 2023-10-20 at 12 36 01 AM" src="https://github.com/TurboWarp/extensions/assets/139097378/ebec3654-6d7a-4bfe-8da1-1695b1612bd8">

Skyhigh173 committed:
more blocks & update & bug fix

Skyhigh173 committed:
remade almost all things

Skyhigh173 committed:
Indexed Object

BlueDome77 commented:
> Update File: [Objects.txt](https://github.com/TurboWarp/extensions/files/13051450/Objects.txt)
> 
> I added 2 new blocks, they are incredibly helpful to me, and what I believe others would find useful too: <img alt="Screenshot 2023-10-20 at 12 35 44 AM" width="361" src="https://user-images.githubusercontent.com/139097378/276857096-9b91315e-59b6-44ce-982e-3be6442f1104.png"> <img alt="Screenshot 2023-10-20 at 12 36 01 AM" width="361" src="https://user-images.githubusercontent.com/139097378/276857101-ebec3654-6d7a-4bfe-8da1-1695b1612bd8.png">
> 
> (Note: I didn't bother adding default values so you can add them if you want...)

Maybe change **placement** to **index**...

Skyhigh173 commented:
> > Update File: [Objects.txt](https://github.com/TurboWarp/extensions/files/13051450/Objects.txt)
> 
> > 
> 
> > I added 2 new blocks, they are incredibly helpful to me, and what I believe others would find useful too: <img alt="Screenshot 2023-10-20 at 12 35 44 AM" width="361" src="https://user-images.githubusercontent.com/139097378/276857096-9b91315e-59b6-44ce-982e-3be6442f1104.png"> <img alt="Screenshot 2023-10-20 at 12 36 01 AM" width="361" src="https://user-images.githubusercontent.com/139097378/276857101-ebec3654-6d7a-4bfe-8da1-1695b1612bd8.png">
> 
> > 
> 
> > (Note: I didn't bother adding default values so you can add them if you want...)
> 
> 
> 
> Maybe change **placement** to **index**...

i have already changed it to *index*

InfinityLoop1 commented:
you can use prettier.io/playground to fix the formatting issue

LilyMakesThings commented:
> you can use prettier.io/playground to fix the formatting issue

please don't comment this on every failed check, formatting is done during merge, it's not required

SharkPool-SP commented:
> > you can use prettier.io/playground to fix the formatting issue
> 
> 
> 
> please don't comment this on every failed check, formatting is done during merge, it's not required

Thank you, was getting annoying

Skyhigh173 commented:
oh no I forgot indexed object concat

Falkimore commented:
it would be nice to have official object compatibility... is this going to get worked on again?

CubesterYT commented:
@Falkimore it's impossible to do so, as we're limited by our types. Best we can do is use the system that SkyHigh is using in this scenario where we only have strings.

CubesterYT reviewed (commented):
Since there is no activity, I will assume it's ready, so I will do formatting and translations, and leave it up to Garbo.

CubesterYT committed:
Merge branch 'master' into pr/Skyhigh173/1066

CubesterYT committed:
Formatting and Translations
